File: python/src/smol/front/ast.py
"""The abstract syntax tree."""

from dataclasses import dataclass
from enum import StrEnum


class BOp(StrEnum):
    MUL = "mul"
    DIV = "div"
    ADD = "add"
    SUB = "sub"
    LT = "lt"


@dataclass(frozen=True, slots=True)
class Var:
    name: str


@dataclass(frozen=True, slots=True)
class Const:
    value: int


@dataclass(frozen=True, slots=True)
class BinOp:
    op: BOp
    lhs: Expr
    rhs: Expr


@dataclass(frozen=True, slots=True)
class Negate:
    inner: Expr


type Expr = Var | Const | BinOp | Negate


@dataclass(frozen=True, slots=True)
class Assign:
    name: str
    value: Expr


@dataclass(frozen=True, slots=True)
class Print:
    value: Expr


@dataclass(frozen=True, slots=True)
class Read:
    name: str


@dataclass(frozen=True, slots=True)
class If:
    guard: Expr
    true_branch: tuple[Stmt, ...]
    false_branch: tuple[Stmt, ...]


type Stmt = Assign | Print | Read | If


@dataclass(frozen=True, slots=True)
class Program:
    statements: tuple[Stmt, ...]

File: java/src/main/java/smol/front/Ast.java
package smol.front;

import java.util.List;

/** The abstract syntax tree. */
public final class Ast {
  private Ast() {}

  public record Program(List<Stmt> stmts) {
    public Program {
      stmts = List.copyOf(stmts);
    }
  }

  public sealed interface Stmt permits Assign, Print, Read, If {}

  public record Assign(String id, Expr expr) implements Stmt {}

  public record Print(Expr expr) implements Stmt {}

  public record Read(String id) implements Stmt {}

  public record If(Expr guard, List<Stmt> tt, List<Stmt> ff) implements Stmt {
    public If {
      tt = List.copyOf(tt);
      ff = List.copyOf(ff);
    }
  }

  public sealed interface Expr permits Var, Const, BinOp, Negate {}

  public record Var(String id) implements Expr {}

  public record Const(long value) implements Expr {}

  public record BinOp(BOp op, Expr lhs, Expr rhs) implements Expr {}

  public record Negate(Expr inner) implements Expr {}

  public enum BOp {
    MUL("mul"),
    DIV("div"),
    ADD("add"),
    SUB("sub"),
    LT("lt");

    private final String text;

    BOp(String text) {
      this.text = text;
    }

    @Override
    public String toString() {
      return text;
    }
  }
}

